src/test/CalldataLogic.t.sol

```diff
@@ -0,0 +1,284 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.10;
+
+import {DataTypes} from './../../contracts/protocol/libraries/types/DataTypes.sol';
+import {ReserveConfiguration} from './../../contracts/protocol/libraries/configuration/ReserveConfiguration.sol';
+
+import {CalldataLogic} from './../../contracts/protocol/libraries/logic/CalldataLogic.sol';
+import {L2Encoder} from './../../contracts/misc/L2Encoder.sol';
+
+import {Vm} from './../Vm.sol';
+import {TestHelper} from './TestHelper.sol';
+
+// TODO: There is an error with the revert messages. For now, we are checking after reverts only,
+// as the integration tests are already catching the revert messages
+contract CalldataLogicTest is TestHelper {
+  using ReserveConfiguration for DataTypes.ReserveConfigurationMap;
+
+  Vm constant VM = Vm(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D);
+
+  mapping(uint256 => address) reservesList;
+
+  function setUp() public {
+    for (uint256 i = 0; i < 128; i++) {
+      reservesList[i] = address(uint160(400 + i));
+    }
+  }
+
+  function testDecodeSupplyParams(
+    uint16 assetId,
+    uint128 amount,
+    uint16 referralCode
+  ) public {
+    VM.assume(assetId < 128);
+
+    bytes32 args;
+    assembly {
+      args := add(assetId, add(shl(16, amount), shl(144, referralCode)))
+    }
+
+    (address _asset, uint256 _amount, uint16 _referralCode) = CalldataLogic.decodeSupplyParams(
+      reservesList,
+      args
+    );
+
+    assertEq(_asset, reservesList[assetId]);
+    assertEq(_amount, amount);
+    assertEq(_referralCode, referralCode);
+  }
+
+  function testDecodeSupplyWithPermitParams(
+    uint16 assetId,
+    uint128 amount,
+    uint16 referralCode,
+    uint256 deadLine,
+    uint8 permitV
+  ) public {
+    VM.assume(assetId < 128 && deadLine < type(uint32).max);
```

Comment:
`assume` should mainly be used for very narrow checks—broad checks will slow down tests as it will take a while to find valid values, and the test may fail if you hit the max number of rejects. (You can also [configure](https://github.com/gakonst/foundry/blob/a00c1939becf3b40f1453e330d3d9507a756d610/config/src/lib.rs#L181-L188) `fuzz-max-local-rejects` and `fuzz-max-global-rejects` in the `foundry.toml` file).

I'd recommend changing this test (and similar for other `assume` usage in the repo) to either:

```solidity
  function testDecodeSupplyWithPermitParams(
    uint8 assetId, // changed from uint16 to uint8 to reduce number of rejects
    uint128 amount,
    uint16 referralCode,
    uint32 deadLine, // changed to uint32 so fuzzer max value for this param is now type(uint32).max
    uint8 permitV
  ) public {
    VM.assume(assetId < 128);
```

OR, when solidity's types aren't a great fit, use a `uint256` and wrap values manually with modulo, either using Solmate's [`bound`](https://github.com/Rari-Capital/solmate/blob/67d907c82f50649f8168061dcfae8617110da361/src/test/utils/DSTestPlus.sol#L116-L135) method or just using modulo directly (you want `uint256` to reduce modulo bias). A downside of this approach is that when a test fails, you need to take the mod of the failed input to see the actual value used in tests

```solidity
  function testDecodeSupplyWithPermitParams(
    uint256 assetId, // changed to uint256
    uint128 amount,
    uint16 referralCode,
    uint256 deadLine, // changed to uint256
    uint8 permitV
  ) public {
    assetId = assetId % 128; // wrap assetId between 0 and 127
    deadLine = deadLine % 2**32; // wraps deadLine between 0 and 2**32 - 1
```